### ISSUE TITLE
Add a Levi-Civita Symbol (parity of permutation) to Base

### ISSUE THREAD

**Issue by pabloferz** [JuliaLang/julia#10172]
As pointed out <a href="https://groups.google.com/forum/#!topic/julia-users/r9hX-HZqflo">here</a>. It would be useful to have a Levi-Civita Symbol implementation in Base.

It's not clear what the name should be, some proposals are `levicivita`, `parity`, `sign`, `signature`, or `sgn`, although `sign(v::AbstractVector)` already returns an element-wise version of `sign`.

Anyway, I wrote three implementations in this <a href="https://gist.github.com/pabloferz/01675f1bf4c8be359767">gist</a>. The `levicivita_inplace_check` in there is the fastest (I think).

Also, I think it might be added in `combinatorics.jl` but I'm not sure.

What do you think?

**Comment by blakejohnson** [comment]
I think if we call it `levicivita` that really has a special meaning of a 3-argument symbol \epsilon_{ijk}. So, I would propose either `parity` or `signature`.

**Comment by blakejohnson** [comment]
I take that back about levicivita being specialized to 3 arguments. Looks like that is just how it appears most often in physics. Wikipedia, at least, is happy to define it generally.

**Comment by blakejohnson** [comment]
I also did some benchmarking of your various implementations. Here's what I see (plotting performance relative to the in-place checking version):

![levi-bench](https://cloud.githubusercontent.com/assets/866695/6170377/3a701438-b2a2-11e4-8ec9-ea8b9cdec11c.png)

**Comment by pabloferz** [comment]
I had `while j ≠ i` instead of `while j ≠ start`, in the in-place algorithm, probably you were timing an error.

How are you calculating the relative performance?

**Comment by blakejohnson** [comment]
Yeah, I caught that error already in the in-place algorithm.

The benchmarking code I used was:
```julia
macro testit(fn, n)
    quote
        t = @elapsed $fn(randperm($n))
        ct = max(ceil(Int, 0.5/t), 1)
        t2 = @elapsed for _ in 1:ct
            $fn(randperm($n))
        end
        t + t2, ct
    end
end

function bench(perms)
    t1, _ = @testit levicivita perms
    t2, _ = @testit levicivita_simple perms
    t3, _ = @testit levicivita_inplace_check perms
    return [t1, t2, t3]
end

results = hcat(map(bench, [2^n for n = 1:10])...)
results ./= results[:,3]
```
Looking at this again, though, I am also measuring the performance of `randperm`, which hides some of the relative differences between the implementations. I should probably factor that out of the loop.

**Comment by pabloferz** [comment]
Yeah, I think you shoud do it like this:

```julia
macro testit(fn, perm)
    quote
        t = @elapsed $fn($perm)
        ct = max(ceil(Int, 0.5/t), 1)
        t2 = @elapsed for _ in 1:ct
            $fn($perm)
        end
        t + t2, ct
    end
end

function bench(perm)
    t1, _ = @testit levicivita perm
    t2, _ = @testit levicivita_simple perm
    t3, _ = @testit levicivita_inplace_check perm
    return [t1, t2, t3]
end

results = hcat(map(bench, [randperm(2^n) for n = 1:10])...)
results ./= results[:,3]
```

**Comment by blakejohnson** [comment]
Wow, actually there are quite a few problems with my benchmark. I also should have divided the time by `ct`.

**Comment by blakejohnson** [comment]
Yeah, that makes a big difference. New `timeit` macro:
```julia
macro testit(fn, n)
    quote
        p = randperm($n)
        t = @elapsed $fn(p)
        ct = max(ceil(Int, 0.5/t), 1)
        perms = [randperm($n) for _ in 1:ct]
        t = @elapsed for i in 1:ct
            $fn(perms[i])
        end
        t/ct, ct
    end
end
```

And corresponding results:
![levi-bench](https://cloud.githubusercontent.com/assets/866695/6175158/482b88e4-b2c2-11e4-8f45-1c2d90088666.png)

**Comment by blakejohnson** [comment]
"zooming in" on the small `n` region isn't very clear. Clearly there is significant gc noise:

![levi-bench-linear](https://cloud.githubusercontent.com/assets/866695/6175249/fe4124cc-b2c2-11e4-9b62-f8c5bd516a01.png)

**Comment by StefanKarpinski** [comment]
Any reason not to use the in-place version?

**Comment by blakejohnson** [comment]
@StefanKarpinski the in-place version is 20-30% slower for really small inputs (e.g. n = 3). But, probably the right approach is to use the in-place version. If you really want this thing in an inner loop for small inputs, you should probably just wrap it in a memo-izer.

**Comment by StefanKarpinski** [comment]
Maybe a hybrid that has the best of both worlds?

**Comment by jiahao** [comment]
(Historical trivia: the earliest reference I could find to the Levi-Civita symbol in the literature is a [physics paper dating to 1955](https://books.google.com/books?id=PLoOAAAAIAAJ&q=%22Levi-Civita+symbol%22&dq=%22Levi-Civita+symbol%22&hl=en&sa=X&ei=gRDdVIOXLoWsyQSh4YCoCQ&ved=0CDUQ6AEwBThu), where the rank 4 version is used and defined:

![books](https://cloud.githubusercontent.com/assets/1732/6176779/87602dfa-b2cf-11e4-9a14-57d949dd0289.png)
)

**Comment by pabloferz** [comment]
Maybe we could define

```const levicivita_lut = cat(3, [0 0 0; 0 0 1; 0 -1 0], [0 0 -1; 0 0 0; 1 0 0], [0 1 0; -1 0 0; 0 0 0])```

and use that for the case of 3D (which is probably the most used) and the in-place for the rest of the cases.

@jiahao: Nice finding, although if it is refered as "the so-called Levi-Civita symbol", it's probably defined as such somewhere else, but I bet it's not easy to find that.

**Comment by pabloferz** [comment]
```julia
const levicivita_lut = cat(3, [0 0 0; 0 0 1; 0 -1 0], [0 0 -1; 0 0 0; 1 0 0], [0 1 0; -1 0 0; 0 0 0])

function levicivita{T::Integer}(p::AbstractVector{T})
    length(p) == 3 && isperm(p) && return levicivita_lut[p...]
    levicivita_inplace(p)
end
```

**Comment by jiahao** [comment]
@pabloferz good luck finding it!

I had Richard Stanley's "Enumerative Combinatorics" available and there he refers to this operation as "the sign of the permutation".

**Comment by pabloferz** [comment]
http://journals.aps.org/pr/abstract/10.1103/PhysRev.74.456

**Comment by pao** [comment]
> isperm(p)

@pabloferz is this check needed for the LUT approach? Non-permutations are just 0s in the LUT. Should save a branch.

**Comment by pabloferz** [comment]
@pao you are right, yo do not need ```isperm(p)```. I put it there for consistency.

```levicivita_inplace``` returns 0 even when the elements of ```p``` do not fall in the range ```0 < p[i] <= n``` and the LUT solution do not cover those cases. We may check if ```0 < sum(p) < 10``` or something like that instead.

**Comment by pabloferz** [comment]
I was thinking that instead of the LUT we might as well just use

```julia
function levicivita{T::Integer}(p::AbstractVector{T})
    length(p) == 3 && 0 < sum(p) <= 9 && sign((p[2]-p[1])*(p[3]-p[1])*(p[3]-p[2]))
    levicivita_inplace(p)
end
```

**Comment by blakejohnson** [comment]
@pabloferz sure, that would probably be as fast.

**Comment by pabloferz** [comment]
I think my in-place function is slightly faster (for large inputs) than that of @toivoh in #8534, although making an in-place version of his implementation might be as fast or even faster. I will have to check it.

About the case of the 3-symbol Levi-Civita tensor, maybe it is worth to have a `levicivita3` function that uses a LUT or the proposed above approach and a general `levicivita` function.

**Comment by pabloferz** [comment]
Ok. I updated the gist with the options I think are the best solution. My proposal would be:

* To have unsafe and safe versions of `levicicita3`, that implement the 3D Levi-Civita symbol.
* To have a general `levicivita` function that checks in place whether or not the vector is a permutation.
* Use `levicivita` to define `parity` which returns `0` if the permutation is even or `1` if it's odd. Then, you can ask `iseven(parity(p))`.

What do you think?

**Comment by blakejohnson** [comment]
Closed via acd1b66.
